tsdb/engine/tsm1/data_file.go

```diff
@@ -0,0 +1,666 @@
+}
+
+// Entry returns the index entry for the specified key and timestamp.  If no entry
+// matches the key an timestamp, nil is returned.
+func (d *indirectIndex) Entry(key string, timestamp time.Time) *IndexEntry {
+	entries := d.Entries(key)
+	for _, entry := range entries {
+		if entry.Contains(timestamp) {
+			return entry
+		}
+	}
+	return nil
+}
+
+// MarshalBinary returns a byte slice encoded version of the index.
+func (d *indirectIndex) MarshalBinary() ([]byte, error) {
+	return d.b, nil
+}
+
+// UnmarshalBinary populates an index from an encoded byte slice
+// representation of an index.
+func (d *indirectIndex) UnmarshalBinary(b []byte) error {
+	// Keep a reference to the actual index bytes
+	d.b = b
+
+	// To create our "indirect" index, we need to find he location of all the keys in
+	// the raw byte slice.  The keys are listed once each (in sorted order).  Following
+	// each key is a time ordered list of index entry blocks for that key.  The loop below
+	// basically skips across the slice keeping track of the counter when we are at a key
+	// field.
+	var i int32
+	for i < int32(len(b)) {
+		d.offsets = append(d.offsets, i)
+		keyLen := int32(btou16(b[i : i+2]))
+		// Skip to the start of the key
+		i += 2
+
+		// Skip over the key
+		i += keyLen
+
+		// Count of all the index blocks for this key
+		count := int32(btou16(b[i : i+2]))
+
+		// Skip the count bytes
+		i += 2
+
+		// Skip over all the blocks
+		i += count * indexEntrySize
+	}
+	return nil
+}
+
+func (d *indirectIndex) readKey(b []byte) (n int, key string, err error) {
+	// 2 byte size of key
+	n, size := 2, int(btou16(b[:2]))
+
+	// N byte key
+	key = string(b[n : n+size])
+	n += len(key)
+	return
+}
+
+func (d *indirectIndex) readEntries(b []byte) (n int, entries indexEntries, err error) {
+	// 2 byte count of index entries
+	n, count := 2, int(btou16(b[:2]))
+
+	for i := 0; i < count; i++ {
+		ie := &IndexEntry{}
+		if err := ie.UnmarshalBinary(b[i*indexEntrySize+2 : i*indexEntrySize+2+indexEntrySize]); err != nil {
+			return 0, nil, fmt.Errorf("readEntries: unmarshal error: %v", err)
+		}
+		entries = append(entries, ie)
+		n += indexEntrySize
+	}
+	return
+}
+
+// tsmWriter writes keys and values in the TSM format
+type tsmWriter struct {
+	w     io.Writer
+	index TSMIndex
+	n     int64
+}
+
+func NewTSMWriter(w io.Writer) (TSMWriter, error) {
+	n, err := w.Write(append(u32tob(MagicNumber), Version))
+	if err != nil {
+		return nil, err
+	}
+
+	index := &directIndex{
+		blocks: map[string]indexEntries{},
+	}
+
+	return &tsmWriter{w: w, index: index, n: int64(n)}, nil
+}
+
+func (t *tsmWriter) Write(key string, values Values) error {
+	block, err := values.Encode(nil)
+	if err != nil {
+		return err
+	}
+
+	checksum := crc32.ChecksumIEEE(block)
+
+	n, err := t.w.Write(append(u32tob(checksum), block...))
+	if err != nil {
+		return err
+	}
+
+	// Record this block in index
+	t.index.Add(key, values[0].Time(), values[len(values)-1].Time(), t.n, uint32(n))
```

Comment:
Yes.  Although it would work for queries, but a file wouldn't actually be available to queries until it has been fully written and closed on disk.